File: starters/flutter/lib/src/core/platform/runtime_contract.dart
import 'dart:async';

const commonRuntimeContractVersion = 'common/2.0.0';
const mobileRuntimeContractVersion = 'mobile/2.0.0';
const telemetryExporterContractVersion = 'telemetry-exporter/2.0.0';
const mobileExtensionContractVersion = 'mobile-extension/2.0.0';

enum RuntimeEnvironment { local, staging, production }

final class RuntimeConfiguration {
  const RuntimeConfiguration({
    required this.environment,
    required this.apiBaseUrl,
    required this.requestTimeout,
  });

  final RuntimeEnvironment environment;
  final Uri apiBaseUrl;
  final Duration requestTimeout;

  static RuntimeConfiguration parse(Map<String, String?> values) {
    final environment = switch (values['APP_ENV'] ?? 'local') {
      'local' => RuntimeEnvironment.local,
      'staging' => RuntimeEnvironment.staging,
      'production' => RuntimeEnvironment.production,
      _ => throw const FormatException(
        'APP_ENV must be local, staging or production.',
      ),
    };
    final apiBaseUrl = Uri.tryParse(
      values['API_BASE_URL'] ?? 'http://localhost:3000',
    );
    if (apiBaseUrl == null ||
        !apiBaseUrl.hasScheme ||
        !apiBaseUrl.hasAuthority) {
      throw const FormatException('API_BASE_URL must be an absolute URI.');
    }
    if (environment == RuntimeEnvironment.production &&
        apiBaseUrl.scheme != 'https') {
      throw const FormatException('Production API endpoints must use HTTPS.');
    }
    final timeoutMs = int.tryParse(values['API_TIMEOUT_MS'] ?? '15000');
    if (timeoutMs == null || timeoutMs < 100 || timeoutMs > 120000) {
      throw const FormatException(
        'API_TIMEOUT_MS must be between 100 and 120000.',
      );
    }
    return RuntimeConfiguration(
      environment: environment,
      apiBaseUrl: apiBaseUrl,
      requestTimeout: Duration(milliseconds: timeoutMs),
    );
  }
}

final class CanonicalMobileError implements Exception {
  const CanonicalMobileError({
    required this.statusCode,
    required this.errorCode,
    required this.message,
    required this.details,
    required this.path,
    required this.timestamp,
    required this.requestId,
  });

  factory CanonicalMobileError.fromJson(Object? value) {
    if (value case {
      'statusCode': final int statusCode,
      'errorCode': final String errorCode,
      'message': final String message,
      'path': final String path,
      'timestamp': final String timestamp,
      'requestId': final String requestId,
    }) {
      final parsedTimestamp = DateTime.tryParse(timestamp);
      if (parsedTimestamp != null) {
        return CanonicalMobileError(
          statusCode: statusCode,
          errorCode: errorCode,
          message: message,
          details: (value as Map<String, Object?>)['details'],
          path: path,
          timestamp: parsedTimestamp.toUtc(),
          requestId: requestId,
        );
      }
    }
    return CanonicalMobileError(
      statusCode: 0,
      errorCode: 'NETWORK_ERROR',
      message: 'The request could not be completed.',
      details: null,
      path: '',
      timestamp: DateTime.fromMillisecondsSinceEpoch(0, isUtc: true),
      requestId: '',
    );
  }

  final int statusCode;
  final String errorCode;
  final String message;
  final Object? details;
  final String path;
  final DateTime timestamp;
  final String requestId;
}

final _traceparent = RegExp(
  r'^00-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$',
);

final class MobileRequestContext {
  const MobileRequestContext({
    required this.requestId,
    required this.traceparent,
  });

  factory MobileRequestContext.create({
    required String requestId,
    required String Function() nextTraceId,
    required String Function() nextSpanId,
    String? incomingTraceparent,
  }) {
    if (requestId.trim().isEmpty) throw ArgumentError('requestId is required.');
    final incoming = incomingTraceparent == null
        ? null
        : _traceparent.firstMatch(incomingTraceparent.toLowerCase());
    final traceId = incoming?.group(1) ?? nextTraceId();
    final spanId = nextSpanId();
    final flags = incoming?.group(3) ?? '01';
    if (!RegExp(r'^[0-9a-f]{32}$').hasMatch(traceId) ||
        !RegExp(r'^[0-9a-f]{16}$').hasMatch(spanId)) {
      throw const FormatException('Invalid W3C trace or span identifier.');
    }
    return MobileRequestContext(
      requestId: requestId,
      traceparent: '00-$traceId-$spanId-$flags',
    );
  }

  final String requestId;
  final String traceparent;
}

const _sensitiveKeys = <String>{
  'authorization',
  'cookie',
  'password',
  'token',
  'accesstoken',
  'refreshtoken',
  'secret',
  'apikey',
  'email',
  'phone',
};

Object? redactRuntimeValue(Object? value) {
  if (value is Map) {
    return <String, Object?>{
      for (final entry in value.entries)
        entry.key.toString():
            _sensitiveKeys.contains(
              entry.key.toString().toLowerCase().replaceAll(
                RegExp('[^a-z0-9]'),
                '',
              ),
            )
            ? '[Redacted]'
            : redactRuntimeValue(entry.value),
    };
  }
  if (value is Iterable) return value.map(redactRuntimeValue).toList();
  if (value is String) {
    return value.replaceAll(
      RegExp(r'Bearer\s+[A-Za-z0-9._~+/-]+', caseSensitive: false),
      'Bearer [Redacted]',
    );
  }
  if (value is Error || value is Exception) return value.runtimeType.toString();
  return value;
}

typedef StructuredSink = void Function(Map<String, Object?> record);

final class StructuredLogger {
  StructuredLogger(this._sink, {DateTime Function()? clock})
    : _clock = clock ?? DateTime.now;

  final StructuredSink _sink;
  final DateTime Function() _clock;

  void log(
    String level,
    String message, {
    MobileRequestContext? context,
    Map<String, Object?> fields = const {},
  }) {
    _sink(
      Map.unmodifiable(
        redactRuntimeValue(<String, Object?>{
              'level': level,
              'message': message,
              'timestamp': _clock().toUtc().toIso8601String(),
              if (context != null) 'requestId': context.requestId,
              if (context != null) 'traceparent': context.traceparent,
              'fields': fields,
            })
            as Map<String, Object?>,
      ),
    );
  }
}

final class TechnicalAudit {
  TechnicalAudit(this._sink, {DateTime Function()? clock})
    : _clock = clock ?? DateTime.now;

  final StructuredSink _sink;
  final DateTime Function() _clock;

  void record(
    String action,
    String outcome,
    MobileRequestContext context, {
    Map<String, Object?> fields = const {},
  }) {
    _sink(
      Map.unmodifiable(
        redactRuntimeValue(<String, Object?>{
              'eventType': 'technical-audit',
              'action': action,
              'outcome': outcome,
              'requestId': context.requestId,
              'traceparent': context.traceparent,
              'timestamp': _clock().toUtc().toIso8601String(),
              'fields': fields,
            })
            as Map<String, Object?>,
      ),
    );
  }
}

abstract interface class MobileTelemetryExporter {
  String get contractVersion;
  void export(Map<String, Object?> record);
}

final class RuntimeTelemetry {
  RuntimeTelemetry(this._exporter, {DateTime Function()? clock})
    : _clock = clock ?? DateTime.now {
    if (_exporter.contractVersion != telemetryExporterContractVersion) {
      throw ArgumentError('Unsupported telemetry exporter contract.');
    }
  }

  final MobileTelemetryExporter _exporter;
  final DateTime Function() _clock;
  final Map<String, int> _counters = {};

  void record(
    String name,
    MobileRequestContext context, {
    Map<String, Object?> fields = const {},
  }) {
    _counters[name] = (_counters[name] ?? 0) + 1;
    _exporter.export(
      Map.unmodifiable(
        redactRuntimeValue(<String, Object?>{
              'name': name,
              'timestamp': _clock().toUtc().toIso8601String(),
              'requestId': context.requestId,
              'traceparent': context.traceparent,
              'fields': fields,
            })
            as Map<String, Object?>,
      ),
    );
  }

  int count(String name) => _counters[name] ?? 0;
}

abstract interface class DiagnosticCheck {
  String get id;
  String run();
}

final class RuntimeDiagnostics {
  final Map<String, DiagnosticCheck> _checks = {};

  void register(DiagnosticCheck check) {
    if (!RegExp(r'^[a-z][a-z0-9.-]*$').hasMatch(check.id) ||
        _checks.containsKey(check.id)) {
      throw ArgumentError('Invalid or duplicate diagnostic check: ${check.id}');
    }
    _checks[check.id] = check;
  }

  List<Map<String, String>> snapshot() {
    final checks = _checks.values.toList()
      ..sort((left, right) => left.id.compareTo(right.id));
    return List.unmodifiable(
      checks.map(
        (check) => Map<String, String>.unmodifiable({
          'id': check.id,
          'status': check.run(),
        }),
      ),
    );
  }
}

abstract interface class RuntimeLifecycleHook {
  String get id;
  FutureOr<void> start();
  FutureOr<void> stop();
}

final class RuntimeLifecycle {
  RuntimeLifecycle(this._hooks);

  final List<RuntimeLifecycleHook> _hooks;
  final List<RuntimeLifecycleHook> _started = [];

  Future<void> start() async {
    if (_started.isNotEmpty) return;
    for (final hook in _hooks) {
      await hook.start();
      _started.add(hook);
    }
  }

  Future<void> stop() async {
    for (final hook in _started.reversed) {
      await hook.stop();
    }
    _started.clear();
  }
}

enum MobileExtensionKind {
  secureStorage,
  session,
  offline,
  push,
  crashReporting,
}

abstract interface class MobileRuntimeExtension {
  MobileExtensionKind get kind;
  String get contractVersion;
  String get id;
}

final class MobileRuntimeExtensionRegistry {
  final Map<MobileExtensionKind, MobileRuntimeExtension> _extensions = {};

  void register(MobileRuntimeExtension extension) {
    if (extension.contractVersion != mobileExtensionContractVersion) {
      throw ArgumentError('Unsupported mobile extension contract.');
    }
    if (_extensions.containsKey(extension.kind)) {
      throw StateError('${extension.kind.name} is already registered.');
    }
    _extensions[extension.kind] = extension;
  }

  MobileRuntimeExtension? get(MobileExtensionKind kind) => _extensions[kind];
}

File: starters/react-native/src/platform/runtime-contract.ts
import { createLogger, type LogRecord, type LogSink } from '../logger/logger';
import { redactValue } from '../logger/redaction';

export const COMMON_RUNTIME_CONTRACT_VERSION = 'common/2.0.0';
export const MOBILE_RUNTIME_CONTRACT_VERSION = 'mobile/2.0.0';
export const TELEMETRY_EXPORTER_CONTRACT_VERSION = 'telemetry-exporter/2.0.0';
export const MOBILE_EXTENSION_CONTRACT_VERSION = 'mobile-extension/2.0.0';

export interface RuntimeConfiguration {
  readonly environment: 'local' | 'staging' | 'production';
  readonly apiBaseUrl: string;
  readonly requestTimeoutMs: number;
}

export function validateRuntimeConfiguration(
  input: Readonly<Record<string, string | undefined>>,
): RuntimeConfiguration {
  const environment = input.EXPO_PUBLIC_APP_ENV ?? 'local';
  if (!['local', 'staging', 'production'].includes(environment)) {
    throw new Error('EXPO_PUBLIC_APP_ENV must be local, staging or production.');
  }
  const apiBaseUrl = new URL(input.EXPO_PUBLIC_API_BASE_URL ?? 'http://localhost:3000');
  if (environment === 'production' && apiBaseUrl.protocol !== 'https:') {
    throw new Error('Production API endpoints must use HTTPS.');
  }
  const requestTimeoutMs = Number(input.EXPO_PUBLIC_API_TIMEOUT_MS ?? '15000');
  if (!Number.isInteger(requestTimeoutMs) || requestTimeoutMs < 100 || requestTimeoutMs > 120_000) {
    throw new Error('EXPO_PUBLIC_API_TIMEOUT_MS must be an integer between 100 and 120000.');
  }
  return Object.freeze({
    environment: environment as RuntimeConfiguration['environment'],
    apiBaseUrl: apiBaseUrl.toString().replace(/\/$/, ''),
    requestTimeoutMs,
  });
}

export interface CanonicalMobileError {
  readonly statusCode: number;
  readonly errorCode: string;
  readonly message: string;
  readonly details: unknown;
  readonly path: string;
  readonly timestamp: string;
  readonly requestId: string;
}

export function mapCanonicalMobileError(value: unknown): CanonicalMobileError {
  const candidate = value as Partial<CanonicalMobileError> | null;
  if (
    !candidate
    || !Number.isInteger(candidate.statusCode)
    || typeof candidate.errorCode !== 'string'
    || typeof candidate.message !== 'string'
    || typeof candidate.path !== 'string'
    || typeof candidate.timestamp !== 'string'
    || typeof candidate.requestId !== 'string'
  ) {
    return {
      statusCode: 0,
      errorCode: 'NETWORK_ERROR',
      message: 'The request could not be completed.',
      details: null,
      path: '',
      timestamp: new Date(0).toISOString(),
      requestId: '',
    };
  }
  return {
    statusCode: candidate.statusCode as number,
    errorCode: candidate.errorCode,
    message: candidate.message,
    details: candidate.details ?? null,
    path: candidate.path,
    timestamp: candidate.timestamp,
    requestId: candidate.requestId,
  };
}

const TRACEPARENT = /^00-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/;

export interface MobileRequestContext {
  readonly requestId: string;
  readonly traceparent: string;
}

export function createMobileRequestContext(options: {
  readonly incomingTraceparent?: string;
  readonly requestId: string;
  readonly nextTraceId: () => string;
  readonly nextSpanId: () => string;
}): MobileRequestContext {
  if (!options.requestId.trim()) throw new Error('requestId is required.');
  const incoming = options.incomingTraceparent?.toLowerCase().match(TRACEPARENT);
  const traceId = incoming?.[1] ?? options.nextTraceId();
  const flags = incoming?.[3] ?? '01';
  const spanId = options.nextSpanId();
  if (!/^[0-9a-f]{32}$/.test(traceId) || !/^[0-9a-f]{16}$/.test(spanId)) {
    throw new Error('Trace and span identifiers must be lowercase hexadecimal W3C identifiers.');
  }
  return Object.freeze({
    requestId: options.requestId,
    traceparent: `00-${traceId}-${spanId}-${flags}`,
  });
}

export type TechnicalAuditSink = (event: Readonly<Record<string, unknown>>) => void;

export class TechnicalAudit {
  constructor(
    private readonly sink: TechnicalAuditSink,
    private readonly clock: () => string = () => new Date().toISOString(),
  ) {}

  record(
    action: string,
    outcome: 'success' | 'failure',
    context: MobileRequestContext,
    fields: Readonly<Record<string, unknown>> = {},
  ): void {
    this.sink(Object.freeze(redactValue({
      eventType: 'technical-audit',
      action,
      outcome,
      requestId: context.requestId,
      traceparent: context.traceparent,
      timestamp: this.clock(),
      fields,
    }) as Record<string, unknown>));
  }
}

export interface MobileTelemetryExporter {
  readonly contractVersion: typeof TELEMETRY_EXPORTER_CONTRACT_VERSION;
  export(record: Readonly<Record<string, unknown>>): void;
}

export class RuntimeTelemetry {
  private readonly counters = new Map<string, number>();

  constructor(
    private readonly exporter: MobileTelemetryExporter,
    private readonly clock: () => string = () => new Date().toISOString(),
  ) {}

  record(name: string, context: MobileRequestContext, fields: Readonly<Record<string, unknown>> = {}): void {
    this.counters.set(name, (this.counters.get(name) ?? 0) + 1);
    this.exporter.export(Object.freeze(redactValue({
      name,
      timestamp: this.clock(),
      requestId: context.requestId,
      traceparent: context.traceparent,
      fields,
    }) as Record<string, unknown>));
  }

  count(name: string): number {
    return this.counters.get(name) ?? 0;
  }
}

export interface DiagnosticCheck {
  readonly id: string;
  run(): 'ready' | 'degraded' | 'unavailable';
}

export class RuntimeDiagnostics {
  private readonly checks = new Map<string, DiagnosticCheck>();

  register(check: DiagnosticCheck): void {
    if (!/^[a-z][a-z0-9.-]*$/.test(check.id) || this.checks.has(check.id)) {
      throw new Error(`Invalid or duplicate diagnostic check: ${check.id}`);
    }
    this.checks.set(check.id, check);
  }

  snapshot(): readonly Readonly<{ id: string; status: string }>[] {
    return [...this.checks.values()]
      .sort((left, right) => left.id.localeCompare(right.id))
      .map((check) => Object.freeze({ id: check.id, status: check.run() }));
  }
}

export interface RuntimeLifecycleHook {
  readonly id: string;
  start(): void | Promise<void>;
  stop(): void | Promise<void>;
}

export class RuntimeLifecycle {
  private started: RuntimeLifecycleHook[] = [];

  constructor(private readonly hooks: readonly RuntimeLifecycleHook[]) {}

  async start(): Promise<void> {
    if (this.started.length) return;
    for (const hook of this.hooks) {
      await hook.start();
      this.started.push(hook);
    }
  }

  async stop(): Promise<void> {
    for (const hook of [...this.started].reverse()) await hook.stop();
    this.started = [];
  }
}

export type MobileExtensionKind =
  | 'secure-storage'
  | 'session'
  | 'offline'
  | 'push'
  | 'crash-reporting';

export interface MobileRuntimeExtension {
  readonly kind: MobileExtensionKind;
  readonly contractVersion: typeof MOBILE_EXTENSION_CONTRACT_VERSION;
  readonly id: string;
}

export class MobileRuntimeExtensionRegistry {
  private readonly extensions = new Map<MobileExtensionKind, MobileRuntimeExtension>();

  register(extension: MobileRuntimeExtension): void {
    if (extension.contractVersion !== MOBILE_EXTENSION_CONTRACT_VERSION) {
      throw new Error(`Unsupported mobile extension contract: ${extension.contractVersion}`);
    }
    if (this.extensions.has(extension.kind)) {
      throw new Error(`A ${extension.kind} extension is already registered.`);
    }
    this.extensions.set(extension.kind, Object.freeze({ ...extension }));
  }

  get(kind: MobileExtensionKind): MobileRuntimeExtension | undefined {
    return this.extensions.get(kind);
  }
}

export function createRuntimeLogger(sink: LogSink): ReturnType<typeof createLogger> {
  return createLogger({ sink, level: 'debug' });
}

export type { LogRecord };
